- complete training with cluster (e.g. work through any wikis or tutorials); 
- get comfortable with how to download from the web to the cluster
	- download the [HCA Immune cells](https://preview.data.humancellatlas.org) and save to cluster
	- read about the [1.3 million neurons from 10X Genomics](https://support.10xgenomics.com/single-cell-gene-expression/datasets) and the [TENxBrainData](https://www.bioconductor.org/packages/devel/data/experiment/vignettes/TENxBrainData/inst/doc/TENxBrainData.html) in Bioconductor, which allows you to import this data in a SingleCellExperiment object. Work through the TENxBrainData tutorial. 
- create a smaller subset of each object by selecting columns or chunks of cells and reading them into R
- filter the rows to remove all genes with zero counts
- work through the [DropletUtils](https://www.bioconductor.org/packages/3.7/bioc/html/DropletUtils.html) vignette to learn how to compute barcode ranks, detect empty droplets and reduct the effect of barcode swapping.
- work through the [scater](https://bioconductor.org/packages/release/bioc/html/scater.html) vignette to learn about how about how to apply quality control metrics
- calculate the total number of counts for each cell to be used a scaling factor for normalization -- also can read about the normalize function in scater
- Download the `ClusterR` package and explore its functions
	- In particular, understand difference between KMeans_arma and KMeans_Rcpp
	- Try KMeans with different initializations: random, kmeans++
